src/uct/base/uct_iface.h

```diff
@@ -406,36 +406,48 @@ typedef struct uct_iface_local_addr_ns {
     void uct_##_name##_cleanup(void);
 
 
+#define _UCT_IFACE_UNQUOTE(_x) _x
+#define _UCT_IFACE_CTOR_
+#define _UCT_IFACE_DTOR_
+#define _UCT_IFACE_CTOR_ctor UCS_F_CTOR
+#define _UCT_IFACE_DTOR_ctor UCS_F_DTOR
+
 /**
- * Register component and TL
+ * Register/unregister TL
  *
- * @param [in] _name   Component and TL name
+ * @param [in] _name          Component and TL name
+ * @param [in] _scope         Scope for functions, must be ctor or empty
+ * @param [in] _init_code     Initialization code
+ * @param [in] _cleanup_code  Cleanup code
  */
-#define UCT_TL_INIT(_name) \
-    void uct_##_name##_init(void); \
-    static void uct_component_tl_##_name##_ctor(void); \
-    void uct_##_name##_init(void) { \
-        uct_component_tl_##_name##_ctor(); \
-        uct_component_##_name##_ctor(); \
+#define UCT_TL_INIT(_name, _scope, _init_code, _cleanup_code) \
+    _UCT_IFACE_UNQUOTE(_UCT_IFACE_CTOR_##_scope) void uct_##_name##_init(void) \
+    { \
+        _init_code; \
         uct_tl_##_name##_ctor(); \
     } \
-    void uct_component_tl_##_name##_ctor(void)
+    _UCT_IFACE_UNQUOTE(_UCT_IFACE_DTOR_##_scope) void uct_##_name##_cleanup(void) \
+    { \
+        uct_tl_##_name##_dtor(); \
```

Comment:
> can we call `uct_tl_register(_component, &uct_##_name##_tl)` here directly instead of defining `uct_tl_##_name##_ctor/dtor` in UCT_TL_DEFINE_ENTRY ?

done